File: searching/fibonacci_search.rs
/**
 * Fibonacci Search Algorithm Implementation
 *
 * Time Complexity: O(log n)
 * Space Complexity: O(1)
 *
 * WHEN TO USE FIBONACCI SEARCH OVER BINARY SEARCH:
 * 1. When division/multiplication operations are costly (embedded systems, old CPUs)
 * 2. For data stored on magnetic tapes or systems where jumping backward is expensive
 * 3. When you want to minimize comparisons on average (fewer than binary search)
 * 4. For uniformly distributed sorted data
 *
 * RUST-SPECIFIC ADVANTAGES:
 * - Overflow-safe arithmetic with checked operations
 * - Pattern matching for cleaner comparison logic
 * - Zero-cost abstractions maintain performance
 */

use std::cmp::min;
use std::time::Instant;

/// Perform Fibonacci search on a sorted array
pub fn fibonacci_search(arr: &[i32], target: i32) -> Option<usize> {
    let n = arr.len();
    if n == 0 {
        return None;
    }

    // Initialize Fibonacci numbers
    let mut fib_m2 = 0;  // (m-2)'th Fibonacci number
    let mut fib_m1 = 1;  // (m-1)'th Fibonacci number
    let mut fib_m = fib_m2 + fib_m1;  // m'th Fibonacci number

    // Find the smallest Fibonacci number >= n
    while fib_m < n {
        fib_m2 = fib_m1;
        fib_m1 = fib_m;
        fib_m = fib_m2 + fib_m1;
    }

    // Marks the eliminated range from front
    let mut offset: isize = -1;

    // While there are elements to be inspected
    while fib_m > 1 {
        // Check if fib_m2 is a valid index
        let i = min((offset + fib_m2 as isize) as usize, n - 1);

        // Use pattern matching for cleaner comparison
        match arr[i].cmp(&target) {
            std::cmp::Ordering::Less => {
                // Target is greater
                fib_m = fib_m1;
                fib_m1 = fib_m2;
                fib_m2 = fib_m - fib_m1;
                offset = i as isize;
            }
            std::cmp::Ordering::Greater => {
                // Target is less
                fib_m = fib_m2;
                fib_m1 = fib_m1 - fib_m2;
                fib_m2 = fib_m - fib_m1;
            }
            std::cmp::Ordering::Equal => {
                // Element found
                return Some(i);
            }
        }
    }

    // Compare the last element
    let last_idx = (offset + 1) as usize;
    if fib_m1 == 1 && last_idx < n && arr[last_idx] == target {
        return Some(last_idx);
    }

    None
}

/// Optimized Fibonacci search with early termination
pub fn fibonacci_search_optimized(arr: &[i32], target: i32) -> Option<usize> {
    let n = arr.len();
    if n == 0 {
        return None;
    }

    // Quick boundary checks
    if target < arr[0] || target > arr[n - 1] {
        return None;
    }
    if arr[0] == target {
        return Some(0);
    }
    if arr[n - 1] == target {
        return Some(n - 1);
    }

    // Initialize Fibonacci numbers
    let mut fib_m2 = 0;
    let mut fib_m1 = 1;
    let mut fib_m = fib_m2 + fib_m1;

    while fib_m < n {
        fib_m2 = fib_m1;
        fib_m1 = fib_m;
        fib_m = fib_m2 + fib_m1;
    }

    let mut offset: isize = -1;

    while fib_m > 1 {
        let i = min((offset + fib_m2 as isize) as usize, n - 1);

        match arr[i].cmp(&target) {
            std::cmp::Ordering::Less => {
                fib_m = fib_m1;
                fib_m1 = fib_m2;
                fib_m2 = fib_m - fib_m1;
                offset = i as isize;
            }
            std::cmp::Ordering::Greater => {
                fib_m = fib_m2;
                fib_m1 = fib_m1 - fib_m2;
                fib_m2 = fib_m - fib_m1;
            }
            std::cmp::Ordering::Equal => return Some(i),
        }
    }

    let last_idx = (offset + 1) as usize;
    if fib_m1 == 1 && last_idx < n && arr[last_idx] == target {
        return Some(last_idx);
    }

    None
}

/// Binary search for comparison
fn binary_search(arr: &[i32], target: i32) -> Option<usize> {
    let mut left = 0;
    let mut right = arr.len();

    while left < right {
        let mid = left + (right - left) / 2;
        match arr[mid].cmp(&target) {
            std::cmp::Ordering::Equal => return Some(mid),
            std::cmp::Ordering::Less => left = mid + 1,
            std::cmp::Ordering::Greater => right = mid,
        }
    }

    None
}

/// Jump search for comparison
fn jump_search(arr: &[i32], target: i32) -> Option<usize> {
    let n = arr.len();
    let jump = (n as f64).sqrt() as usize;
    let mut prev = 0;

    while prev < n && arr[prev.saturating_add(jump).min(n - 1)] < target {
        prev += jump;
    }

    for i in prev.saturating_sub(jump)..prev.saturating_add(jump).min(n) {
        if arr[i] == target {
            return Some(i);
        }
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_fibonacci_search() {
        let arr = vec![1, 3, 5, 7, 9, 11, 13, 15, 17, 19];
        assert_eq!(fibonacci_search(&arr, 7), Some(3));
        assert_eq!(fibonacci_search(&arr, 20), None);
        assert_eq!(fibonacci_search(&arr, 1), Some(0));
        assert_eq!(fibonacci_search(&arr, 19), Some(9));
    }

    #[test]
    fn test_fibonacci_search_optimized() {
        let arr = vec![1, 3, 5, 7, 9, 11, 13, 15, 17, 19];
        assert_eq!(fibonacci_search_optimized(&arr, 7), Some(3));
        assert_eq!(fibonacci_search_optimized(&arr, 20), None);
    }
}

fn main() {
    // Test correctness
    let test_arr = vec![1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29];
    println!("Test Array: {:?}", test_arr);
    println!("Fibonacci Search for 15: {:?}", fibonacci_search(&test_arr, 15));
    println!("Fibonacci Search for 20: {:?}", fibonacci_search(&test_arr, 20));
    println!("Fibonacci Search for 1: {:?}", fibonacci_search(&test_arr, 1));
    println!("Fibonacci Search for 29: {:?}", fibonacci_search(&test_arr, 29));

    // Performance comparison
    println!("\n--- Performance Comparison ---");
    let sizes = vec![1000, 10000, 100000, 1000000];

    for size in sizes {
        let arr: Vec<i32> = (0..size).map(|i| i * 2).collect();
        let target = arr[size / 2];

        // Warm-up
        for _ in 0..100 {
            fibonacci_search(&arr, target);
            binary_search(&arr, target);
            jump_search(&arr, target);
        }

        // Fibonacci search
        let start = Instant::now();
        for _ in 0..10000 {
            fibonacci_search(&arr, target);
        }
        let fib_time = start.elapsed();

        // Binary search
        let start = Instant::now();
        for _ in 0..10000 {
            binary_search(&arr, target);
        }
        let binary_time = start.elapsed();

        // Jump search
        let start = Instant::now();
        for _ in 0..10000 {
            jump_search(&arr, target);
        }
        let jump_time = start.elapsed();

        println!("\nArray size: {}", size);
        println!("Fibonacci Search: {:.3}ms", fib_time.as_secs_f64() * 1000.0);
        println!("Binary Search: {:.3}ms", binary_time.as_secs_f64() * 1000.0);
        println!("Jump Search: {:.3}ms", jump_time.as_secs_f64() * 1000.0);
        println!("Ratio (Fib/Binary): {:.2}x",
                 fib_time.as_secs_f64() / binary_time.as_secs_f64());
        println!("Ratio (Fib/Jump): {:.2}x",
                 fib_time.as_secs_f64() / jump_time.as_secs_f64());
    }

    // Performance on different data distributions
    println!("\n--- Performance on Different Data Distributions ---");

    // Uniform distribution
    let arr_uniform: Vec<i32> = (0..100000).collect();
    let target1 = 75000;

    let start = Instant::now();
    for _ in 0..10000 {
        fibonacci_search(&arr_uniform, target1);
    }
    let uniform_time = start.elapsed();
    println!("Uniform distribution: {:.3}ms", uniform_time.as_secs_f64() * 1000.0);

    // Sparse distribution
    let arr_sparse: Vec<i32> = (0..10000).map(|i| i * 10).collect();
    let target2 = 75000;

    let start = Instant::now();
    for _ in 0..10000 {
        fibonacci_search(&arr_sparse, target2);
    }
    let sparse_time = start.elapsed();
    println!("Sparse distribution: {:.3}ms", sparse_time.as_secs_f64() * 1000.0);
}

File: searching/fibonacci_search.py
"""
Fibonacci Search Algorithm Implementation

Time Complexity: O(log n)
Space Complexity: O(1)

WHEN TO USE FIBONACCI SEARCH OVER BINARY SEARCH:
1. When division/multiplication operations are costly (embedded systems, old CPUs)
2. For data stored on magnetic tapes or systems where jumping backward is expensive
3. When you want to minimize comparisons on average (fewer than binary search)
4. For uniformly distributed sorted data

PERFORMANCE CHARACTERISTICS:
- Uses Fibonacci numbers to divide the array (golden ratio divisions)
- Only uses addition and subtraction (no division or multiplication)
- Average case: slightly fewer comparisons than binary search
- Works well with sequential access patterns
- Better cache performance than binary search in some scenarios

ADVANTAGES OVER BINARY SEARCH:
- Avoids expensive division operations
- Only moves forward (good for tape storage, linked lists)
- Golden ratio division can be more optimal for uniformly distributed data
- Better locality of reference in some cases

ADVANTAGES OVER JUMP SEARCH:
- Logarithmic time complexity O(log n) vs O(√n)
- More efficient for large datasets
- Adaptive block sizes based on Fibonacci sequence
"""

from typing import List


def fibonacci_search(arr: List[int], target: int) -> int:
    """
    Perform Fibonacci search on a sorted array.

    Args:
        arr: Sorted list of integers
        target: Value to search for

    Returns:
        Index of target if found, -1 otherwise
    """
    n = len(arr)
    if n == 0:
        return -1

    # Initialize Fibonacci numbers
    fib_m2 = 0  # (m-2)'th Fibonacci number
    fib_m1 = 1  # (m-1)'th Fibonacci number
    fib_m = fib_m2 + fib_m1  # m'th Fibonacci number

    # Find the smallest Fibonacci number >= n
    while fib_m < n:
        fib_m2 = fib_m1
        fib_m1 = fib_m
        fib_m = fib_m2 + fib_m1

    # Marks the eliminated range from front
    offset = -1

    # While there are elements to be inspected
    while fib_m > 1:
        # Check if fib_m2 is a valid index
        i = min(offset + fib_m2, n - 1)

        # If target is greater than the value at index fib_m2,
        # cut the subarray from offset to i
        if arr[i] < target:
            fib_m = fib_m1
            fib_m1 = fib_m2
            fib_m2 = fib_m - fib_m1
            offset = i

        # If target is less than the value at index fib_m2,
        # cut the subarray after i+1
        elif arr[i] > target:
            fib_m = fib_m2
            fib_m1 = fib_m1 - fib_m2
            fib_m2 = fib_m - fib_m1

        # Element found
        else:
            return i

    # Compare the last element
    if fib_m1 and offset + 1 < n and arr[offset + 1] == target:
        return offset + 1

    return -1


def fibonacci_search_iterative(arr: List[int], target: int) -> int:
    """
    Alternative Fibonacci search with clearer iteration logic.

    Args:
        arr: Sorted list of integers
        target: Value to search for

    Returns:
        Index of target if found, -1 otherwise
    """
    n = len(arr)
    if n == 0:
        return -1

    # Generate Fibonacci numbers up to n
    fibs = [0, 1]
    while fibs[-1] < n:
        fibs.append(fibs[-1] + fibs[-2])

    # Start with the largest Fibonacci number <= n
    k = len(fibs) - 1
    offset = 0

    while k > 0:
        # Calculate the index to check
        idx = min(offset + fibs[k - 1] - 1, n - 1)

        if arr[idx] == target:
            return idx
        elif arr[idx] < target:
            # Move offset forward
            offset = idx + 1
            k -= 1
        else:
            # Reduce the Fibonacci index
            k -= 2

        # Check if we've exhausted the search space
        if offset >= n:
            break

    return -1


def fibonacci_search_optimized(arr: List[int], target: int) -> int:
    """
    Optimized Fibonacci search with early termination and boundary checks.

    Args:
        arr: Sorted list of integers
        target: Value to search for

    Returns:
        Index of target if found, -1 otherwise
    """
    n = len(arr)
    if n == 0:
        return -1

    # Quick boundary checks
    if target < arr[0] or target > arr[-1]:
        return -1

    if arr[0] == target:
        return 0
    if arr[-1] == target:
        return n - 1

    # Generate Fibonacci numbers
    fib_m2, fib_m1 = 0, 1
    fib_m = fib_m2 + fib_m1

    while fib_m < n:
        fib_m2 = fib_m1
        fib_m1 = fib_m
        fib_m = fib_m2 + fib_m1

    offset = -1

    while fib_m > 1:
        i = min(offset + fib_m2, n - 1)

        if arr[i] < target:
            fib_m = fib_m1
            fib_m1 = fib_m2
            fib_m2 = fib_m - fib_m1
            offset = i
        elif arr[i] > target:
            fib_m = fib_m2
            fib_m1 = fib_m1 - fib_m2
            fib_m2 = fib_m - fib_m1
        else:
            return i

    if fib_m1 and offset + 1 < n and arr[offset + 1] == target:
        return offset + 1

    return -1


# Performance testing and demonstration
if __name__ == "__main__":
    import time
    import random

    # Test correctness
    test_arr = [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29]
    print("Test Array:", test_arr)
    print(f"Fibonacci Search for 15: Index {fibonacci_search(test_arr, 15)}")
    print(f"Fibonacci Search for 20: Index {fibonacci_search(test_arr, 20)}")
    print(f"Fibonacci Search for 1: Index {fibonacci_search(test_arr, 1)}")
    print(f"Fibonacci Search for 29: Index {fibonacci_search(test_arr, 29)}")

    # Performance comparison
    print("\n--- Performance Comparison ---")
    sizes = [1000, 10000, 100000, 1000000]

    for size in sizes:
        arr = sorted(random.sample(range(size * 10), size))
        target = random.choice(arr)

        # Fibonacci search
        start = time.perf_counter()
        for _ in range(1000):
            fibonacci_search(arr, target)
        fib_time = time.perf_counter() - start

        # Binary search for comparison
        start = time.perf_counter()
        for _ in range(1000):
            import bisect
            bisect.bisect_left(arr, target)
        binary_time = time.perf_counter() - start

        # Jump search for comparison
        import math
        def jump_search(arr, target):
            """Simple jump-search baseline so timings are comparable."""
            n = len(arr)
            jump = int(math.sqrt(n))
            prev = 0
            while prev < n and arr[min(prev + jump, n - 1)] < target:
                prev += jump
            for i in range(max(0, prev - jump), min(prev + jump, n)):
                if arr[i] == target:
                    return i
            return -1

        start = time.perf_counter()
        for _ in range(1000):
            jump_search(arr, target)
        jump_time = time.perf_counter() - start

        print(f"\nArray size: {size:,}")
        print(f"Fibonacci Search: {fib_time*1000:.3f}ms")
        print(f"Binary Search: {binary_time*1000:.3f}ms")
        print(f"Jump Search: {jump_time*1000:.3f}ms")
        print(f"Ratio (Fib/Binary): {fib_time/binary_time:.2f}x")
        print(f"Ratio (Fib/Jump): {fib_time/jump_time:.2f}x")

    # Comparison count analysis
    print("\n--- Comparison Count Analysis ---")
    print("Fibonacci search typically makes fewer comparisons than binary search")
    print("for uniformly distributed data due to golden ratio divisions.")

    # Test on different data distributions
    print("\n--- Performance on Different Data Distributions ---")

    # Uniform distribution
    arr_uniform = list(range(100000))
    target = 75000

    start = time.perf_counter()
    for _ in range(10000):
        fibonacci_search(arr_uniform, target)
    uniform_time = time.perf_counter() - start

    print(f"Uniform distribution: {uniform_time*1000:.3f}ms")

    # Sparse distribution (gaps between elements)
    arr_sparse = [i * 10 for i in range(10000)]
    target = 750000

    start = time.perf_counter()
    for _ in range(10000):
        fibonacci_search(arr_sparse, target)
    sparse_time = time.perf_counter() - start

    print(f"Sparse distribution: {sparse_time*1000:.3f}ms")
